pkg/pillar/zedbox/zedbox.go

```diff
@@ -33,46 +40,227 @@ import (
+		"conntrack":        conntrack.Run,
+		"tpmmgr":           tpmmgr.Run,
+		"vaultmgr":         vaultmgr.Run,
+		"upgradeconverter": upgradeconverter.Run,
+	}
+	log *base.LogObject
+)
 
 func main() {
-	pid := os.Getpid()
-	basename := filepath.Base(os.Args[0])
-	log := base.NewSourceLogObject(basename, pid)
-	ps := pubsub.New(&socketdriver.SocketDriver{Log: log}, log)
-	if entrypoint, ok := entrypoints[basename]; ok {
-		entrypoint(ps)
+	versionPtr := flag.Bool("v", false, "Version")
+	debugPtr := flag.Bool("d", false, "Debug flag")
+	debugOverride = *debugPtr
+	if *versionPtr {
+		fmt.Printf("%s: %s\n", os.Args[0], Version)
+		return
+	}
+
+	log = agentlog.Init(agentName)
+
+	if debugOverride {
+		logrus.SetLevel(logrus.DebugLevel)
 	} else {
-		fmt.Printf("Unknown package: %s\n", basename)
+		logrus.SetLevel(logrus.InfoLevel)
+	}
+
+	// Check what service we are intending to start.
+	basename := filepath.Base(os.Args[0])
+	if _, ok := entrypoints[basename]; ok {
+		// If its a known child service, the notify zedbox binary to start that
+		sericeInitStatus := types.ServiceInitStatus{
+			ServiceName: basename,
+			CmdArgs:     os.Args,
+		}
+		log.Infof("Notifying zedbox to start service %s with args %v",
+			sericeInitStatus.ServiceName, sericeInitStatus.CmdArgs)
+		if err := publish(agentName, &sericeInitStatus); err != nil {
+			log.Fatalf(err.Error())
+		}
+		return
+	} else if basename != agentName { // If its unknown child service, check if we intend to start zedbox
+		fmt.Printf("zedbox: Unknown package: %s\n", basename)
 		os.Exit(1)
 	}
+
+	//Start zedbox
+	var sktData string
+	sktChan := make(chan string)
+	stillRunning := time.NewTicker(15 * time.Second)
+	pid := os.Getpid()
+	logObj := base.NewSourceLogObject(agentName, pid)
+	ps := pubsub.New(&socketdriver.SocketDriver{Log: logObj}, logObj)
+
+	log.Infof("Starting %s", agentName)
+	if err := pidfile.CheckAndCreatePidfile(log, agentName); err != nil {
+		log.Fatal(err)
+	}
+
+	go startSubscriber(agentName, types.ServiceInitStatus{}, sktChan)
+	for {
+		select {
+		case sktData = <-sktChan:
+			sktData = strings.TrimSpace(sktData)
+			var serviceInitStatus types.ServiceInitStatus
+			if err := json.Unmarshal([]byte(sktData), &serviceInitStatus); err != nil {
+				err := fmt.Errorf("zedbox: exception while unmarshalling data %s. %s",
+					sktData, err.Error())
+				log.Errorf(err.Error())
+				break
+			}
+
+			log.Infof("zedbox: Received command = %s agrs = %v",
+				serviceInitStatus.ServiceName, serviceInitStatus.CmdArgs)
+			srvLogObj := base.NewSourceLogObject(serviceInitStatus.ServiceName, pid)
+			srvPs := pubsub.New(&socketdriver.SocketDriver{Log: srvLogObj}, srvLogObj)
+			if entrypoint, ok := entrypoints[serviceInitStatus.ServiceName]; ok {
+				log.Infof("zedbox: Starting %s", serviceInitStatus.ServiceName)
+				flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+				os.Args = serviceInitStatus.CmdArgs
+				go entrypoint(srvPs)
```

Comment:
Instead of invoking it directly, why not invoke a wrapper which writes agentName.done?
Something like
func runAndDone(agentName string, ps *pubsub.Pubsub) {
     entrypoint(ps)
     log.Infof("service %s done", agentName)
     writeDoneFile(agentnane)
}

That way you avoid code in any of the agents for the done part, and you don't need to define specific filenames in the types package.

Reply:
Changed.